build hkl.3001

prjemian opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

prjemian committed:
MNT another configure option

prjemian committed:
MNT don't patch yet

prjemian committed:
MNT change source repo, no patch

prjemian commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 3.23.3, conda-smithy 3.23.0, and conda-forge-pinning 2023.03.18.10.09.36

prjemian committed:
MNT re-enable patch

prjemian commented:
Need to update the patch

```py

Hunk #1 succeeded at 164 with fuzz 2 (offset 22 lines).
Hunk #2 FAILED at 210.
1 out of 2 hunks FAILED
Warning: failed to download source.  If building, will try again after downloading recipe dependencies.
Error was: 
Command '['/opt/conda/bin/patch', '--no-backup-if-mismatch', '--batch', '-Np1', '-i', '/tmp/tmph0ibinfn/no-docs.patch.native', '--binary', '--dry-run']' returned non-zero exit status 1.
Traceback (most recent call last):
  File "/opt/conda/bin/conda-mambabuild", line 10, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.10/site-packages/boa/cli/mambabuild.py", line 256, in main
    call_conda_build(action, config)
  File "/opt/conda/lib/python3.10/site-packages/boa/cli/mambabuild.py", line 228, in call_conda_build
    result = api.build(
  File "/opt/conda/lib/python3.10/site-packages/conda_build/api.py", line 180, in build
    return build_tree(
  File "/opt/conda/lib/python3.10/site-packages/conda_build/build.py", line 3097, in build_tree
    packages_from_this = build(metadata, stats,
  File "/opt/conda/lib/python3.10/site-packages/conda_build/build.py", line 2136, in build
    try_download(m, no_download_source=False, raise_error=True)
  File "/opt/conda/lib/python3.10/site-packages/conda_build/render.py", line 656, in try_download
    raise RuntimeError("Failed to download or patch source. Please see build log for info.")
RuntimeError: Failed to download or patch source. Please see build log for info.
##[error]Bash exited with code '1'.
Finishing: Run docker build
```

prjemian committed:
MNT #13 revise the patch

prjemian commented:
In the patch:

```patch
-dnl *** gtk-doc ***
-GTK_DOC_CHECK([1.9],[--flavour no-tmpl])
```

This patch could remain but the `--disable-gtk-doc` option to `configure` makes this unnecessary.

```patch
@@ -213,12 +210,6 @@ AC_CONFIG_FILES([Makefile
 		 tests/Makefile
 		 tests/bindings/Makefile
 		 tests/tap/Makefile
-		 Documentation/Makefile
-		 Documentation/hkl.org
-		 Documentation/api/Makefile
-		 Documentation/figures/Makefile
-		 Documentation/sphinx/Makefile
-		 Documentation/sphinx/source/conf.py
 		 gui/Makefile
 		 data/Makefile
 		 contrib/Makefile
```

This code has relocated and changed.  This patch should be re-generated.

prjemian committed:
MNT #13 rebuild patch file

prjemian committed:
MNT #13 build needs HDF5

prjemian commented:
Still this CI error:

```
2023-03-20T14:47:19.7950589Z checking for hdf5 >= 1.8.13... no
2023-03-20T14:47:19.8296234Z configure: error: Package requirements (hdf5 >= 1.8.13) were not met:
2023-03-20T14:47:19.8296803Z 
2023-03-20T14:47:19.8297564Z No package 'hdf5' found
2023-03-20T14:47:19.8297773Z 
2023-03-20T14:47:19.8298169Z Consider adjusting the PKG_CONFIG_PATH environment variable if you
2023-03-20T14:47:19.8298690Z installed software in a non-standard prefix.
2023-03-20T14:47:19.8298935Z 
2023-03-20T14:47:19.8299312Z Alternatively, you may set the environment variables HDF5_CFLAGS
2023-03-20T14:47:19.8299827Z and HDF5_LIBS to avoid the need to call pkg-config.
2023-03-20T14:47:19.8300316Z See the pkg-config man page for more details.
```

Not sure how to resolve.

prjemian committed:
CI #13 apply pattern from boost-cpp feedstock

prjemian commented:
Didn't work as expected:

```
Mamba failed to solve:
 - make
 - pkg-config
 - libtool
 - gcc_linux-64 11.*
 - gxx_linux-64 11.*
 - autoconf
 - m4
 - automake
 - libhdf5-dev

with channels:

The reported errors are:
- Encountered problems while solving:
-   - nothing provides requested libhdf5-dev
```

prjemian commented:
The way past this point is to install `hdf5` into the conda build environment, then define:

```bash
    export HDF5_CFLAGS=" -I${CONDA_PREFIX}/include" \
    export HDF5_LIBS=" -L${CONDA_PREFIX}/lib -lhdf5" \
```

prjemian committed:
MNT #13 configure HDF5

prjemian commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 3.24.0, conda-smithy 3.23.1, and conda-forge-pinning 2023.03.28.18.53.36

prjemian commented:
This faliure:

```bash
/bin/bash ../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I..    -DG_LOG_DOMAIN=\"Hkl\" -I.. -I.. -I../third-party/ -ftrack-macro-expansion=0  -I/root/miniconda/include -I/root/miniconda/include -I/root/miniconda/include/glib-2.0 -I/root/miniconda/lib/glib-2.0/include -I/root/miniconda/include -Wall -g -O2 -MT hkl-engine-template.lo -MD -MP -MF $depbase.Tpo -c -o hkl-engine-template.lo hkl-engine-template.c &&\
mv -f $depbase.Tpo $depbase.Plo
/bin/bash: line 1: --template: command not found
make[3]: *** [Makefile:1552: stamp-hkl-type-builtins.h] Error 127
```

is due to these lines in `hkl/Makefile`:

```make
stamp-hkl-type-builtins.h: $(hkl_public_h_sources) $(srcdir)/hkl-type-builtins.h.template
	( cd $(srcdir) && $(GLIB_MKENUMS) --template hkl-type-builtins.h.template \
```

Specifically, the `GLIB_MKENUMS` symbol is not defined, where it is expected to be found at `$CONDA_PREFIX/bin/glib-mkenums`.

prjemian commented:
Latest failure: <pre><b>ModuleNotFoundError:</b> No module named 'giscanner._giscanner'</b></pre>

```bash
2023-03-28T21:21:55.5414492Z libtool: link: ( cd ".libs" && rm -f "libhkl.la" && ln -s "../libhkl.la" "libhkl.la" )
2023-03-28T21:21:55.5463192Z CPPFLAGS="-DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem $PREFIX/include" CFLAGS="-Wall -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/hkl-5.0.0.3001 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix" LDFLAGS="-Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,$PREFIX/lib -Wl,-rpath-link,$PREFIX/lib -L$PREFIX/lib" CC="$BUILD_PREFIX/bin/x86_64-conda-linux-gnu-cc" PKG_CONFIG="$BUILD_PREFIX/bin/x86_64-conda-linux-gnu-pkg-config" GI_HOST_OS="" DLLTOOL="false"  $PREFIX/bin/g-ir-scanner  --add-include-path=. --add-include-path=. --warn-all --namespace=Hkl --nsversion=5.0 --libtool="/bin/sh ../libtool" --pkg=gobject-2.0  --pkg-export=hkl-5.0   --library=libhkl.la --warn-all --c-include='hkl.h' --cflags-begin   -DG_LOG_DOMAIN=\"Hkl\" -I.. -I.. -I../third-party/ -ftrack-macro-expansion=0 -I$PREFIX/include/glib-2.0 -I$PREFIX/lib/glib-2.0/include -I$PREFIX/include -I$BUILD_PREFIX/include -I$PREFIX/include -I$PREFIX/include/glib-2.0 -I$PREFIX/lib/glib-2.0/include -I$PREFIX/include --cflags-end  ../hkl.h hkl-binding-private.h hkl-types.h hkl-vector.c hkl-matrix.c hkl-quaternion.c hkl-parameter.c hkl-axis.c hkl-geometry.c hkl-detector.c hkl-detector-factory.c hkl-lattice.c hkl-sample.c hkl-pseudoaxis.c hkl-factory.c hkl-binding.c hkl-types.c hkl-type-builtins.c hkl-type-builtins.h $PREFIX/bin/g-ir-scanner libhkl.la Makefile --output Hkl-5.0.gir
2023-03-28T21:21:55.6462585Z Traceback (most recent call last):
2023-03-28T21:21:55.6464146Z   File "/home/conda/feedstock_root/build_artifacts/hkl_1680038274201/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/bin/g-ir-scanner", line 98, in <module>
2023-03-28T21:21:55.6464811Z     from giscanner.scannermain import scanner_main
2023-03-28T21:21:55.6465744Z   File "/home/conda/feedstock_root/build_artifacts/hkl_1680038274201/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/gobject-introspection/giscanner/scannermain.py", line 35, in <module>
2023-03-28T21:21:55.6466683Z     from giscanner.ast import Include, Namespace
2023-03-28T21:21:55.6467593Z   File "/home/conda/feedstock_root/build_artifacts/hkl_1680038274201/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/gobject-introspection/giscanner/ast.py", line 29, in <module>
2023-03-28T21:21:55.6468757Z     from .sourcescanner import CTYPE_TYPEDEF, CSYMBOL_TYPE_TYPEDEF
2023-03-28T21:21:55.6469908Z   File "/home/conda/feedstock_root/build_artifacts/hkl_1680038274201/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/gobject-introspection/giscanner/sourcescanner.py", line 34, in <module>
2023-03-28T21:21:55.6470842Z     from giscanner._giscanner import SourceScanner as CSourceScanner
2023-03-28T21:21:55.6471366Z ModuleNotFoundError: No module named 'giscanner._giscanner'
2023-03-28T21:21:55.6602054Z make[3]: *** [/home/conda/feedstock_root/build_artifacts/hkl_1680038274201/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/share/gobject-introspection-1.0/Makefile.introspection:156: Hkl-5.0.gir] Error 1
2023-03-28T21:21:55.6608979Z make[3]: Leaving directory '$SRC_DIR/hkl'
2023-03-28T21:21:55.6611669Z make[2]: *** [Makefile:1181: all-recursive] Error 1
2023-03-28T21:21:55.6612463Z make[2]: Leaving directory '$SRC_DIR/hkl'
2023-03-28T21:21:55.6626350Z make[1]: *** [Makefile:566: all-recursive] Error 1
2023-03-28T21:21:55.6627513Z make[1]: Leaving directory '$SRC_DIR'
2023-03-28T21:21:55.6630835Z make: *** [Makefile:450: all] Error 2
2023-03-28T21:21:56.0569448Z Traceback (most recent call last):
2023-03-28T21:21:56.0570394Z   File "/opt/conda/bin/conda-mambabuild", line 10, in <module>
2023-03-28T21:21:56.0579287Z     sys.exit(main())
2023-03-28T21:21:56.0580297Z   File "/opt/conda/lib/python3.10/site-packages/boa/cli/mambabuild.py", line 256, in main
2023-03-28T21:21:56.0580827Z     call_conda_build(action, config)
2023-03-28T21:21:56.0581241Z   File "/opt/conda/lib/python3.10/site-packages/boa/cli/mambabuild.py", line 228, in call_conda_build
2023-03-28T21:21:56.0582379Z     result = api.build(
2023-03-28T21:21:56.0583098Z   File "/opt/conda/lib/python3.10/site-packages/conda_build/api.py", line 180, in build
2023-03-28T21:21:56.0592526Z     return build_tree(
2023-03-28T21:21:56.0593755Z   File "/opt/conda/lib/python3.10/site-packages/conda_build/build.py", line 3078, in build_tree
2023-03-28T21:21:56.0615761Z     packages_from_this = build(metadata, stats,
2023-03-28T21:21:56.0616399Z   File "/opt/conda/lib/python3.10/site-packages/conda_build/build.py", line 2198, in build
2023-03-28T21:21:56.0627506Z     utils.check_call_env(cmd, env=env, rewrite_stdout_env=rewrite_env,
2023-03-28T21:21:56.0629961Z   File "/opt/conda/lib/python3.10/site-packages/conda_build/utils.py", line 451, in check_call_env
2023-03-28T21:21:56.0648182Z     return _func_defaulting_env_to_os_environ("call", *popenargs, **kwargs)
2023-03-28T21:21:56.0649142Z   File "/opt/conda/lib/python3.10/site-packages/conda_build/utils.py", line 427, in _func_defaulting_env_to_os_environ
2023-03-28T21:21:56.0649664Z     raise subprocess.CalledProcessError(proc.returncode, _args)
2023-03-28T21:21:56.0650416Z subprocess.CalledProcessError: Command '['/bin/bash', '-o', 'errexit', '/home/conda/feedstock_root/build_artifacts/hkl_1680038274201/work/conda_build.sh']' returned non-zero exit status 2.
2023-03-28T21:22:01.6447863Z ##[error]Bash exited with code '1'.
2023-03-28T21:22:01.6495529Z ##[section]Finishing: Run docker build
```

prjemian committed:
MNT #13 diagnostic

prjemian commented:
In [some other projects](https://duckduckgo.com/?q=ModuleNotFoundError%3A+No+module+named+%27giscanner._giscanner%27&t=brave&ia=web), this error has been resolved by `conda deactivate`.  Only because `gobject-introspection` was installed there into the system Python.  Not the case here.

mrakitin commented:
Power-cycler to rerun the CI jobs.

prjemian commented:
Failed with `No module named 'giscanner._giscanner'`:

```bash
libtool: link: ( cd ".libs" && rm -f "libhkl.la" && ln -s "../libhkl.la" "libhkl.la" )
CPPFLAGS="-DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem $PREFIX/include" CFLAGS="-Wall -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/hkl-5.0.0.3001 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix" LDFLAGS="-Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,$PREFIX/lib -Wl,-rpath-link,$PREFIX/lib -L$PREFIX/lib" CC="$BUILD_PREFIX/bin/x86_64-conda-linux-gnu-cc" PKG_CONFIG="$BUILD_PREFIX/bin/x86_64-conda-linux-gnu-pkg-config" GI_HOST_OS="" DLLTOOL="false"  $PREFIX/bin/g-ir-scanner  --add-include-path=. --add-include-path=. --warn-all --namespace=Hkl --nsversion=5.0 --libtool="/bin/sh ../libtool" --pkg=gobject-2.0  --pkg-export=hkl-5.0   --library=libhkl.la --warn-all --c-include='hkl.h' --cflags-begin   -DG_LOG_DOMAIN=\"Hkl\" -I.. -I.. -I../third-party/ -ftrack-macro-expansion=0 -I$PREFIX/include/glib-2.0 -I$PREFIX/lib/glib-2.0/include -I$PREFIX/include -I$BUILD_PREFIX/include -I$PREFIX/include -I$PREFIX/include/glib-2.0 -I$PREFIX/lib/glib-2.0/include -I$PREFIX/include --cflags-end  ../hkl.h hkl-binding-private.h hkl-types.h hkl-vector.c hkl-matrix.c hkl-quaternion.c hkl-parameter.c hkl-axis.c hkl-geometry.c hkl-detector.c hkl-detector-factory.c hkl-lattice.c hkl-sample.c hkl-pseudoaxis.c hkl-factory.c hkl-binding.c hkl-types.c hkl-type-builtins.c hkl-type-builtins.h $PREFIX/bin/g-ir-scanner libhkl.la Makefile --output Hkl-5.0.gir
Traceback (most recent call last):
  File "/home/conda/feedstock_root/build_artifacts/hkl_1695218021189/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/bin/g-ir-scanner", line 103, in <module>
    from giscanner.scannermain import scanner_main
  File "/home/conda/feedstock_root/build_artifacts/hkl_1695218021189/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/gobject-introspection/giscanner/scannermain.py", line 35, in <module>
    from giscanner.ast import Include, Namespace
  File "/home/conda/feedstock_root/build_artifacts/hkl_1695218021189/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/gobject-introspection/giscanner/ast.py", line 29, in <module>
    from .sourcescanner import CTYPE_TYPEDEF, CSYMBOL_TYPE_TYPEDEF
  File "/home/conda/feedstock_root/build_artifacts/hkl_1695218021189/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/gobject-introspection/giscanner/sourcescanner.py", line 34, in <module>
    from giscanner._giscanner import SourceScanner as CSourceScanner
ModuleNotFoundError: No module named 'giscanner._giscanner'
make[3]: *** [/home/conda/feedstock_root/build_artifacts/hkl_1695218021189/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/share/gobject-introspection-1.0/Makefile.introspection:156: Hkl-5.0.gir] Error 1
make[3]: Leaving directory '$SRC_DIR/hkl'
```

prjemian commented:
Same error as before.

prjemian commented:
Related? https://github.com/AyatanaIndicators/ayatana-ido/issues/56

prjemian commented:
Ayatana issue not related.

prjemian commented:
Most solutions point to version issues in conda when building with gobject-introspection and suggest deactivating conda.  But here, we need conda to build the conda package.  Try these diagnostics in `build.sh` after resolving a merge conflict:

```sh
# diagnostics
conda list obj
echo "g-ir-scanner location: $(which g-ir-scanner)"
echo "g-ir-scanner version: $(g-ir-scanner --version)"
```

prjemian committed:
build diagnostics

prjemian committed:
MNT rebuild meta.yaml

prjemian commented:
@conda-forge-admin, please rerender

prjemian committed:
STY syntax

Commit:
MNT: Re-rendered with conda-build 3.26.1, conda-smithy 3.26.2, and conda-forge-pinning 2023.09.26.15.03.47

prjemian committed:
PKG adjust the pinning

prjemian committed:
PKG pin python

prjemian committed:
PKG

prjemian committed:
pin

prjemian committed:
SHA

prjemian committed:
use git

prjemian committed:
extract tar file

prjemian committed:
syntax

prjemian committed:
diagnostics

prjemian committed:
python test code

prjemian committed:
pin and test

prjemian committed:
move test code

prjemian committed:
new libhkl build

prjemian committed:
re-run the CI

prjemian committed:
re-run

prjemian committed:
revert a couple requirements

prjemian committed:
ease up on testing

prjemian committed:
again

prjemian committed:
updated binaries and requirements

prjemian commented:
@mrakitin Success!  Still in draft while I prepare the repo that provides the libhkl binary file.  "Prepare" means add documentation, instructions, and build files, then move that repo to the bluesky organization.

prjemian commented:
Now builds properly on py38-py311 with recent versions (v5.0.0.3001 and v5.0.0.3357) of libhkl.

prjemian committed:
use public repo as source now

prjemian commented:
@mrakitin This is ready for review!

prjemian commented:
@klauer:  Finally, an update.

klauer commented:
Nicely done @prjemian !

mrakitin commented on recipe/meta.yaml:
```diff
@@ -1,50 +1,44 @@
-  sha256: {{ sha256 }}
-  patches:
-    - no-docs.patch
+  git_rev: main
+  git_url: https://github.com/{{ gitorg }}/libhkl.git
```
We should use GH releases' artifacts for that in the future. I opened an issue to document it: https://github.com/bluesky/libhkl/issues/3.

prjemian replied:
Agreed, [as noted](https://github.com/bluesky/libhkl/issues/3#issuecomment-1741089648).

mrakitin commented on recipe/build.sh:
```diff
@@ -2,14 +2,18 @@
-./configure --disable-static --disable-gui --enable-introspection=yes --disable-hkl-doc --prefix=$PREFIX
-make -j ${CPU_COUNT}
-make install
+TARBALL="libhkl-${PKG_VERSION}-x86_64.tar.gz"
+echo "Installing from ${TARBALL}"
+tar xzf "${TARBALL}" -C "${PREFIX}"
```
So, the approach is to not build it here with conda, but use the pre-built archive?
I wonder if the approach may have some consequences if the toolset in conda-forge changes. Any thoughts?

prjemian replied:
I believe we can respond more easily if the toolset in conda-forge changes.  Other C library packages in conda-forge also build a tarball outside of the recipe, as done here.

prjemian replied:
The way we respond is to build those changes into the tarball process, then re-release a new tarball as needed.

mrakitin reviewed (commented):
It's great the build works now! I wonder why it's not possible to compile it directly here. Thoughts?

prjemian commented:
> I wonder why it's not possible to compile it directly here. Thoughts?

Conflicting libraries, as noted in the comments both above and https://github.com/bluesky/hklpy/issues/250.

prjemian commented:
What took some time was to realize that the library build process had to be separated from conda-forge tools **but** it must use identical (or at least compatible) versions of required packages for the conda-forge recipe.  The specific versions to be matched are pinned in the conda-forge recipe and built from source in the library build process.

prjemian commented:
You might note that the test file [`try_libhkl.py`](https://github.com/bluesky/libhkl/blob/main/tests/try_libhkl.py) is shared (obtained) from the *libhkl* repo.

prjemian commented:
[CF policies](https://github.com/conda-forge/staged-recipes/blob/main/.github/pull_request_template.md) say to use tarballs rather than `git_url`.  Moving this PR back to draft status, awaiting https://github.com/bluesky/libhkl/issues/3.

mrakitin commented on recipe/meta.yaml:
```diff
@@ -1,50 +1,44 @@
 test:
   commands:
+    - ls -lAFgh ${PREFIX}/lib/libhkl.*
     - test -f ${PREFIX}/lib/libhkl.so
```
We'd like to also see how it resolves the dependency libraries:
```suggestion
    - test -f ${PREFIX}/lib/libhkl.so
    - ldd ${PREFIX}/lib/libhkl.so
```

mrakitin committed:
Add `ldd` test for `libhkl.so`

prjemian committed:
MNT use release artefact

prjemian committed:
MNT "sha256", by name

prjemian committed:
MNT try with py312

prjemian commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 3.27.0, conda-smithy 3.26.3, and conda-forge-pinning 2023.10.02.11.43.13

prjemian committed:
CI re-run

prjemian committed:
MNT re-run with newest tag

prjemian committed:
Merge branch 'conda-forge:main' into hkl.3001

prjemian committed:
force CI to re-run

prjemian commented:
Pushing to the branch for this PR but CI does not run again.

prjemian committed:
MNT now v5.0.0.3001, the target of this PR

prjemian committed:
MNT no "v" in the version here, as before

prjemian commented:
@mrakitin, @klauer : ready for review!

klauer reviewed (approved):
The proof is in the pudding, so they say, and this works. Big 👍 from me

mrakitin commented:
Looks like only the linter job was running. What happened with the actual packaging jobs?

mrakitin commented:
https://github.com/conda-forge/hkl-feedstock/pull/13/commits/33a496d541fea4f8f878a2ec424fcea8c6585ac0 - this commit removed a bunch of relevant CI files. I think it should be recovered.

prjemian commented:
I'm [puzzled](https://github.com/conda-forge/hkl-feedstock/pull/13#issuecomment-1743494776), too.

prjemian commented:
@conda-forge-admin, please rerender

prjemian commented:
No outages on [GitHub](https://www.githubstatus.com/) or [Azure](https://azure.status.microsoft/en-us/status) status.

mrakitin commented on recipe/meta.yaml:
```diff
@@ -1,50 +1,48 @@
-  number: 7
-  skip: True  # [py<36 or win or osx]
+  number: 0
+  skip: True  # [py<38 or py>12 or win or osx]
```
```suggestion
  skip: True  # [py<38 or py>312 or win or osx]
```

prjemian commented:
Or global service issues: https://portal.azure.com/#view/Microsoft_Azure_Health/AzureHealthBrowseBlade/~/serviceIssues

mrakitin commented:
I think it's the "skip" condition:
```yaml
skip: True  # [py<38 or py>12 or win or osx]
```
I suggested the fix.

prjemian committed:
Update recipe/meta.yaml

Aha!  That's good.  It's an upgrade suggested today by the autobot, includes my typo.

Co-authored-by: Max Rakitin <mrakitin@users.noreply.github.com>

mrakitin commented:
@conda-forge-admin, please rerender.

Commit:
MNT: Re-rendered with conda-build 3.27.0, conda-smithy 3.26.3, and conda-forge-pinning 2023.10.02.17.25.16

prjemian commented:
Failed here:

> tar (child): libhkl-5.0.0.3001-x86_64.tar.gz: Cannot open: No such file or directory


I'll revert this one: [c251a30](https://github.com/conda-forge/hkl-feedstock/pull/13/commits/c251a30248fb4932a9c712130b33f2670a32aee5)

mrakitin commented on recipe/build.sh:
```diff
@@ -2,14 +2,18 @@
-./configure --disable-static --disable-gui --enable-introspection=yes --disable-hkl-doc --prefix=$PREFIX
-make -j ${CPU_COUNT}
-make install
+TARBALL="libhkl-${PKG_VERSION}-x86_64.tar.gz"
```
Need to use the same version as downloaded.
```suggestion
TARBALL="libhkl-v${PKG_VERSION}-x86_64.tar.gz"
```

prjemian committed:
Update recipe/build.sh

Co-authored-by: Max Rakitin <mrakitin@users.noreply.github.com>

prjemian commented:
I like that change better.  Keeps the version naming consistent.

prjemian commented:
Not seeing a build for py312 now.  Due to upstream package requirements (gsl and/or gobject-introspection).

prjemian commented:
Still:

> tar (child): libhkl-v5.0.0.3001-x86_64.tar.gz: Cannot open: No such file or directory

mrakitin commented:
The file name matches now, but I think we should use a different directory from where to unpack it. Some info from the logs:
```
Source cache directory is: /home/conda/feedstock_root/build_artifacts/src_cache
Downloading source to cache: libhkl-v5.0.0.3001-x86_64_b0a7941785.tar.gz
Downloading https://github.com/bluesky/libhkl/releases/download/v1.1.0/libhkl-v5.0.0.3001-x86_64.tar.gz
INFO:conda_build.source:Source cache directory is: /home/conda/feedstock_root/build_artifacts/src_cache
INFO:conda_build.source:Downloading source to cache: libhkl-v5.0.0.3001-x86_64_b0a7941785.tar.gz
INFO:conda_build.source:Downloading https://github.com/bluesky/libhkl/releases/download/v1.1.0/libhkl-v5.0.0.3001-x86_64.tar.gz
Success

```

prjemian commented:
But which directory?

```
2023-10-02T20:07:24.3370164Z INFO:conda_build.source:Source cache directory is: /home/conda/feedstock_root/build_artifacts/src_cache
2023-10-02T20:07:24.3371523Z INFO:conda_build.source:Downloading source to cache: libhkl-v5.0.0.3001-x86_64_b0a7941785.tar.gz
```

This part: `b0a7941785` is the first 10 char of the SHA256.  Is it the `/home/conda/feedstock_root/build_artifacts/src_cache` dir?  I'll run a check.

prjemian committed:
Diagnostic

prjemian committed:
space

mrakitin commented:
I think so. Looks like `conda-build` also attempts to unpack the archive, but we do it separately in `build.sh`.

mrakitin commented:
Interesting, yes, it adds 10 symbols of the sha256 hash. Can we use `tar -xvf libhkl-v5.0.0.3001-x86_64_*.tar.gz`?

prjemian commented:
Found it.  I can make a soft link to the downloaded file.

prjemian commented:
This:

```bash
SRC_CACHE=/home/conda/feedstock_root/build_artifacts/src_cache
# TARBALL="libhkl-v${PKG_VERSION}-x86_64.tar.gz"
TARBALL="${SRC_CACHE}/$(ls ${SRC_CACHE} | grep libhkl-*.tar.gz)"
```

prjemian committed:
Find the tarball

prjemian committed:
MNT syntax

prjemian commented:
Can't glob (`*`) in a grep.

prjemian commented:
> 2023-10-02T20:46:03.2085564Z ValueError: License file given in about/license_file (/home/conda/recipe_root/COPYING) does not exist in source root dir or in recipe root dir (with meta.yaml)

I can fix this, tomorrow.

prjemian committed:
license file from hkl

prjemian committed:
add test code to this recipe

prjemian committed:
around for the right directory

prjemian commented:
What is different about this run that it fails (hkl-feedstock (linux linux_64_python3.9.____73_pypy) )?
![image](https://github.com/conda-forge/hkl-feedstock/assets/2279984/c24fe2d7-2d01-4862-a893-f68f28fe3d47)

It fails somewhere in the python test code.  Will add diagnostic text to each `assert` and re-run.

prjemian committed:
add more output

prjemian commented:
```
2023-10-02T21:30:00.8741984Z + test -f /tmp/tmpok0m6ssf/info/recipe/try_libhkl.py
2023-10-02T21:30:00.8742406Z + python /tmp/tmpok0m6ssf/info/recipe/try_libhkl.py
2023-10-02T21:30:00.9644208Z RPython traceback:
2023-10-02T21:30:00.9645748Z   File "pypy_module_cpyext_2.c", line 31937, in type_realize
2023-10-02T21:30:00.9646503Z   File "pypy_module_cpyext_2.c", line 42685, in _type_realize
2023-10-02T21:30:00.9647555Z   File "pypy_module_cpyext_1.c", line 43477, in from_ref
2023-10-02T21:30:00.9647965Z Fatal RPython error: AssertionError
2023-10-02T21:30:01.3640862Z /home/conda/feedstock_root/build_artifacts/hkl_1696281987518/test_tmp/run_test.sh: line 15:   345 Aborted                 (core dumped) python ${RECIPE_DIR}/try_libhkl.py
2023-10-02T21:30:02.0982649Z Tests failed for hkl-5.0.0.3001-py39h4162558_0.conda - moving package to /home/conda/feedstock_root/build_artifacts/broken
```

prjemian committed:
TST Which test is failing?

prjemian commented:
Moving to draft mode since one of the workflows fails.

prjemian committed:
TST earlier

prjemian committed:
TST restart, keep it simple

prjemian committed:
TST now the imports

prjemian committed:
TST just the one import now

prjemian committed:
TST can't import gi?

prjemian committed:
TST confirm by not importing gi

prjemian committed:
TST revert test code to full

prjemian commented:
### Summary

No "_tests_" were failing.  It was the `import gi` line that failed, yet the workflow log showed the libgobject library was installed:

```
2023-10-03T14:43:34.6927933Z 	libgobject-2.0.so.0 => $PREFIX/lib/./libgobject-2.0.so.0 (0x00007f0e28fce000)
```

The logs do not show clearly what, precisely, failed.

prjemian committed:
TST show details when assert fails

klauer commented:
@prjemian correct me if I'm wrong, but it appears that just the PyPy build is failing. Do we care about supporting it for now? I'd imagine anyone who uses hklpy will be after CPython support.

I've found that in other non-pure Python feedstocks PyPy tends to be problematic as well, so you're not alone here.

prjemian commented:
Coming to the same conclusion.  I'm revising the test now to relax the extra asserts I just added.

prjemian committed:
TST one less assert

prjemian commented:
Before I completely surrender on this, I'll ask AI for some assist.

prjemian committed:
TST pivot to run_test.sh

prjemian committed:
TST reference run_test.sh

prjemian committed:
TST indentation

prjemian committed:
TST discard this idea

mrakitin commented:
I think https://github.com/conda-forge/hkl-feedstock/pull/13/commits/cf7a562868d44daf22d8db4799d4fc5d462e5208 made sense. We needed to list the source files, per https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#source-files.

prjemian commented:
That URL suggests the test be refactored into the style of `pytest` (or other recognized framework).  This branch has a lot of content devoted to _just get the tests right_.  That could be worked up in a revision build (new branch and PR) for this same libhkl release (v5.0.0.3001).

mrakitin commented:
I second @klauer's suggestion to exclude the pypy builds for this project.

mrakitin commented:
Lovely!
```
+ test -f /tmp/tmp931x0ico/info/recipe/try_libhkl.py
+ python /tmp/tmp931x0ico/info/recipe/try_libhkl.py
1 E4CH  <Hkl.Factory object at 0x7f36527db5e0 (HklFactory at 0x7f3651a283a0)>
2 E4CV  <Hkl.Factory object at 0x7f36527db220 (HklFactory at 0x7f3651a28460)>
3 E6C  <Hkl.Factory object at 0x7f3652941960 (HklFactory at 0x7f3651a28ea0)>
4 K4CV  <Hkl.Factory object at 0x7f36527db640 (HklFactory at 0x7f3651a28740)>
5 K6C  <Hkl.Factory object at 0x7f36527db400 (HklFactory at 0x7f3651a29ec0)>
6 PETRA3 P09 EH2  <Hkl.Factory object at 0x7f36527db580 (HklFactory at 0x7f3651a2a4c0)>
7 PETRA3 P23 4C  <Hkl.Factory object at 0x7f36527db1c0 (HklFactory at 0x7f3651a2a760)>
8 PETRA3 P23 6C  <Hkl.Factory object at 0x7f36527db460 (HklFactory at 0x7f3651a2aee0)>
9 SOLEIL MARS  <Hkl.Factory object at 0x7f36527db6a0 (HklFactory at 0x7f3651a28400)>
10 SOLEIL NANOSCOPIUM ROBOT  <Hkl.Factory object at 0x7f36527db280 (HklFactory at 0x7f3651a2b100)>
11 SOLEIL SIRIUS KAPPA  <Hkl.Factory object at 0x7f36527db520 (HklFactory at 0x7f3651a29e40)>
12 SOLEIL SIRIUS TURRET  <Hkl.Factory object at 0x7f36527db160 (HklFactory at 0x7f3651a2b2c0)>
13 SOLEIL SIXS MED1+2  <Hkl.Factory object at 0x7f36527db2e0 (HklFactory at 0x7f3651a2b9c0)>
14 SOLEIL SIXS MED2+2  <Hkl.Factory object at 0x7f36527db4c0 (HklFactory at 0x7f3651a2ba20)>
15 SOLEIL SIXS MED2+3  <Hkl.Factory object at 0x7f36527db100 (HklFactory at 0x7f3651a2b940)>
16 SOLEIL SIXS MED2+3 v2  <Hkl.Factory object at 0x7f36527db340 (HklFactory at 0x7f3651a2b8c0)>
17 TwoC  <Hkl.Factory object at 0x7f36529434c0 (HklFactory at 0x7f3651a280c0)>
18 ZAXIS  <Hkl.Factory object at 0x7f36527db3a0 (HklFactory at 0x7f3651a2bbe0)>
+ echo 'libhkl test finished'
+ exit 0
libhkl test finished
```

prjemian commented:
As @klauer said, CI is passing on all the workflows that matter to me.  This one problem workflow has become a distraction since insufficient diagnostics are available to identify the root problem for this one workflow.

prjemian commented:
> I second @klauer's suggestion to exclude the pypy builds for this project.

So we just ignore the failing workflow or is there some action we take?

```
2023-10-03T15:38:14.8492748Z ##[section]Starting: linux linux_64_python3.9.____73_pypy
```

prjemian commented:
For example: https://github.com/conda-forge/staged-recipes/issues/20694

prjemian committed:
TST skip pypy build

prjemian commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 3.27.0, conda-smithy 3.26.3, and conda-forge-pinning 2023.10.03.16.12.15

prjemian commented:
BTW, found that way to exclude PyPY from [a web search](https://duckduckgo.com/?q=How+to+disable+pypy+workflows+in+conda-forge+builds&t=brave&ia=web).

prjemian commented:
All **_enabled_** workflows have succeeded.  Ready for review.  Again.

mrakitin commented on recipe/meta.yaml:
```diff
@@ -1,50 +1,48 @@
-  skip: True  # [py<36 or win or osx]
+  number: 0
+  skip: true  # [python_impl == 'pypy']
+  skip: True  # [py<38 or py>312 or win or osx]
```
Let's use consistent `true` in the `skip` conditions.
```suggestion
  skip: true  # [py<38 or py>312 or win or osx]
```

klauer replied:
I would have expected this to be in a single line with the `python_impl == 'pypy' or` but apparently either way works

klauer reviewed (approved):
OK, now this _actually_ runs on the CI and works. Exciting!

mrakitin reviewed (approved):
Looks good to me, thanks for pushing on that, @prjemian!

I made a minor non-blocking suggestion below.

prjemian committed:
Update recipe/meta.yaml

Co-authored-by: Max Rakitin <mrakitin@users.noreply.github.com>

prjemian commented:
@strempfer: These geometries will be part of the support with the new library version (code changes will be necessary for hklpy to expose them, so that upgrade will happen soon):

```
2023-10-03T17:26:25.6820047Z 1 E4CH  <Hkl.Factory object at 0x7f5341f43d70 (HklFactory at 0x7f53411b53a0)>
2023-10-03T17:26:25.6821197Z 2 E4CV  <Hkl.Factory object at 0x7f5341f439b0 (HklFactory at 0x7f53411b5460)>
2023-10-03T17:26:25.6821887Z 3 E6C  <Hkl.Factory object at 0x7f5341f43650 (HklFactory at 0x7f53411b5ea0)>
2023-10-03T17:26:25.6822670Z 4 K4CV  <Hkl.Factory object at 0x7f5341f43dd0 (HklFactory at 0x7f53411b5740)>
2023-10-03T17:26:25.6823181Z 5 K6C  <Hkl.Factory object at 0x7f5341f43b90 (HklFactory at 0x7f53411b6ec0)>
2023-10-03T17:26:25.6823460Z 6 PETRA3 P09 EH2  <Hkl.Factory object at 0x7f5341f43d10 (HklFactory at 0x7f53411b74c0)>
2023-10-03T17:26:25.6823802Z 7 PETRA3 P23 4C  <Hkl.Factory object at 0x7f5341f43950 (HklFactory at 0x7f53411b7760)>
2023-10-03T17:26:25.6824101Z 8 PETRA3 P23 6C  <Hkl.Factory object at 0x7f5341f43bf0 (HklFactory at 0x7f53411b7ee0)>
2023-10-03T17:26:25.6824377Z 9 SOLEIL MARS  <Hkl.Factory object at 0x7f5341f43e30 (HklFactory at 0x7f53411b5400)>
2023-10-03T17:26:25.6824694Z 10 SOLEIL NANOSCOPIUM ROBOT  <Hkl.Factory object at 0x7f5341f43a10 (HklFactory at 0x7f53411b8100)>
2023-10-03T17:26:25.6825014Z 11 SOLEIL SIRIUS KAPPA  <Hkl.Factory object at 0x7f5341f43cb0 (HklFactory at 0x7f53411b6e40)>
2023-10-03T17:26:25.6825325Z 12 SOLEIL SIRIUS TURRET  <Hkl.Factory object at 0x7f5341f438f0 (HklFactory at 0x7f53411b82c0)>
2023-10-03T17:26:25.6825632Z 13 SOLEIL SIXS MED1+2  <Hkl.Factory object at 0x7f5341f43a70 (HklFactory at 0x7f53411b89c0)>
2023-10-03T17:26:25.6825919Z 14 SOLEIL SIXS MED2+2  <Hkl.Factory object at 0x7f5341f43c50 (HklFactory at 0x7f53411b8a20)>
2023-10-03T17:26:25.6826221Z 15 SOLEIL SIXS MED2+3  <Hkl.Factory object at 0x7f5341f43890 (HklFactory at 0x7f53411b8940)>
2023-10-03T17:26:25.6826509Z 16 SOLEIL SIXS MED2+3 v2  <Hkl.Factory object at 0x7f5341f43ad0 (HklFactory at 0x7f53411b88c0)>
2023-10-03T17:26:25.6826804Z 17 TwoC  <Hkl.Factory object at 0x7f5341f431d0 (HklFactory at 0x7f53411b50c0)>
2023-10-03T17:26:25.6827088Z 18 ZAXIS  <Hkl.Factory object at 0x7f5341f43b30 (HklFactory at 0x7f53411b8be0)>
```
